docs/rfcs/rfc-025-cypher-field-filters.md

```diff
@@ -0,0 +1,302 @@
+# Cypher field filters
+
+## Problem
+
+Currently, `@cypher` fields are evaluated in the `RETURN` statement using `apoc.cypher.runFirstColumnSingle`.
+
+For this reason, it's not possible to use these fields during filtering
+some good reasons why this is needed could be found in the issues:
+
+-   [#1997](https://github.com/neo4j/graphql/issues/1997).
+-   [#554](https://github.com/neo4j/graphql/issues/554).
+
+For instance, given the following typeDefs:
+
+```graphql
+type Author {
+    name: String
+    mostRecentBooks: [Book!]!
+        @cypher(statement: "MATCH (this)-[:AUTHORED_BOOK]->(b:Book) RETURN b AS result ORDER BY b.year DESC LIMIT 5", columnName: "result")
+    lastPublishedYear: Int
+        @cypher(statement: "MATCH (this)-[:AUTHORED_BOOK]->(b:Book) RETURN b.year AS result ORDER BY b.year DESC LIMIT 1", columnName: "result")
+    books: [Book!]! @relationship(type: "AUTHORED_BOOK", direction: OUT)
+}
+
+type Book {
+    name: String!
+    year: Int
+    refID: ID @id
+    soldCopies: @cypher(statement: "OPTIONAL MATCH(sales:Sales) WHERE this.refID = sales.refID WITH count(sales) AS result  RETURN result AS result ", columnName: "result")
+    authors: [Author!]! @relationship(type: "AUTHORED_BOOK", direction: IN)
+}
+
+type Sales {
+    refID: ID
+    price: Int
+}
+```
+
+The following GraphQL query:
+
+```graphql
+query Authors {
+    authors {
+        lastPublishedYear
+    }
+}
+```
+
+will generate a Cypher Query similar to:
+
+```cypher
+MATCH (this:`Author`)
+RETURN this {
+  lastPublishedYear: apoc.cypher.runFirstColumnSingle("MATCH (this)-[:AUTHORED_BOOK]->(b:Book) RETURN b.year ORDER BY b.year DESC LIMIT 1",
+  { this: this })
+} as this
+```
+
+The current library behavior makes it not possible to filter an Author based on the property `lastPublishedYear` as the property is evaluated only in the `RETURN` statement.
+
+## Experimental Solution
+
+This solution relies on the **rfc-028-cypher-directive-subquery** alternative `columnName` solution,
```

Comment:
Just to let you know that `apoc.runFirstColumn` is not necessarily a hard limitation on filtering, as we can filter on a `WITH` statement:

```cypher
MATCH(n:Movie)
WITH n {.title, lastPublished: apoc.cypher.runFirstColumnSingle("RETURN 'test'",{})} as result
WHERE result.lastPublished = "test"
RETURN result
```

Of course, ideally, we would have subqueries as it will be more performant, but this may be doable with both cases, as long as the `runFirstColumn` is done in a subquery and not in the return